dbms/src/Storages/Transaction/FastAddPeer.cpp

```diff
@@ -0,0 +1,498 @@
+#include <Storages/Page/V3/Universal/UniversalPageIdFormatImpl.h>
+#include <Storages/Page/V3/Universal/UniversalPageStorage.h>
+#include <Storages/Page/V3/Universal/UniversalWriteBatchImpl.h>
+#include <Storages/S3/CheckpointManifestS3Set.h>
+#include <Storages/S3/S3Common.h>
+#include <Storages/S3/S3Filename.h>
+#include <Storages/S3/S3GCManager.h>
+#include <Storages/S3/S3RandomAccessFile.h>
+#include <Storages/Transaction/CheckpointInfo.h>
+#include <Storages/Transaction/FastAddPeer.h>
+#include <Storages/Transaction/KVStore.h>
+#include <Storages/Transaction/ProxyFFI.h>
+#include <Storages/Transaction/ProxyFFICommon.h>
+#include <Storages/Transaction/Region.h>
+#include <Storages/Transaction/TMTContext.h>
+#include <fmt/core.h>
+
+namespace DB
+{
+struct AsyncTasks
+{
+    using Key = uint64_t;
+    using Func = std::function<FastAddPeerRes()>;
+
+    explicit AsyncTasks(uint64_t pool_size)
+    {
+        // We use a big queue to cache, to reduce ass task failures.
+        thread_pool = std::make_unique<ThreadPool>(pool_size, pool_size, 300);
+    }
+    explicit AsyncTasks(uint64_t pool_size, uint64_t free_pool_size, uint64_t queue_size)
+    {
+        thread_pool = std::make_unique<ThreadPool>(pool_size, free_pool_size, queue_size);
+    }
+    bool addTask(Key k, Func f)
+    {
+        using P = std::packaged_task<FastAddPeerRes()>;
+        std::shared_ptr<P> p = std::make_shared<P>(P(f));
+
+        auto res = thread_pool->trySchedule([p]() { (*p)(); }, 0, 0);
+        if (res)
+        {
+            std::scoped_lock l(mtx);
+            futures[k] = p->get_future();
+        }
+        return res;
+    }
+
+    bool isScheduled(Key key) const
+    {
+        std::scoped_lock l(mtx);
+        return futures.count(key);
+    }
+
+    bool isReady(Key key) const
+    {
+        using namespace std::chrono_literals;
+        std::scoped_lock l(mtx);
+        if (!futures.count(key))
+            return false;
+        return futures.at(key).wait_for(0ms) == std::future_status::ready;
+    }
+
+    FastAddPeerRes fetchResult(Key key)
+    {
+        std::unique_lock<std::mutex> l(mtx);
+        auto it = futures.find(key);
+        auto fut = std::move(it->second);
+        futures.erase(it);
+        l.unlock();
+        return fut.get();
+    }
+
+protected:
+    std::map<Key, std::future<FastAddPeerRes>> futures;
+    std::unique_ptr<ThreadPool> thread_pool;
+    mutable std::mutex mtx;
+};
+
+FastAddPeerContext::FastAddPeerContext(uint64_t thread_count)
+{
+    if (thread_count == 0)
+    {
+        static constexpr int ffi_handle_sec = 5;
+        static constexpr int region_per_sec = 2;
+        thread_count = ffi_handle_sec * region_per_sec;
+    }
+    tasks_trace = std::make_shared<AsyncTasks>(thread_count);
+}
+
+TempUniversalPageStoragePtr FastAddPeerContext::getTempUniversalPageStorage(UInt64 store_id, const std::string_view & manifest_sub_path)
+{
+    std::unique_lock lock(mu);
```

Comment:
Will there be the situation where a region which requires the older PageStorage version overrides the cached PageStorage with a newer version.

Reply:
This should not be a problem. Before the region is removed, it should always be in the latest manifest.

Reply:
I mean the old PageStorage may overrides the new PageStorage

Reply:
OK, I will fix it.

Reply:
Fixed.